### PR TITLE
Implement the meta/stats API endpoint.

### PR DESCRIPTION
**Pull request by frankban**
As a pre-req, also implement the first stats hook (archive-download).

### CODE REVIEW

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results: http://ci.jujugui.org:8080/job/charmstore/146/

File: params/stats.go [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+	StatsArchiveUpload       = "archive-upload"
+	StatsInfo                = "info"
+	StatsMissing             = "missing"
+)
+
+// Statistic holds one element of a stats/counter
+// response. See http://tinyurl.com/nkdovcf
+type Statistic struct {
+	Key   string `json:",omitempty"`
+	Date  string `json:",omitempty"`
```
Comment (urosj) [inline params/stats.go:23]: The date is holding the date of last change, right? Nice.

File: internal/v4/stats.go [hunk 2]
```diff
@@ -5,15 +5,34 @@ package v4
...
+	"gopkg.in/juju/charm.v3"
 
 	"github.com/juju/charmstore/internal/charmstore"
 	"github.com/juju/charmstore/params"
 )
 
+// entityStatsKey returns a stats key for the given charm or bundle
+// reference and the given kind.
+// Entities' stats can then be retrieved like the following:
+//   - kind:utopic:* -> all charm of a specific series;
```
Comment (urosj) [inline internal/v4/stats.go]: plural in: all charms of a ...

**Comment by jrwren** [comment]
:+1:

File: internal/v4/api.go [hunk 3]
```diff
@@ -307,8 +308,20 @@ func (h *handler) metaArchiveSize(entity *mongodoc.Entity, id *charm.Reference,
 
 // GET id/meta/stats/
 // http://tinyurl.com/lvyp2l5
-func (h *handler) metaStats(id *charm.Reference, path string, method string, flags url.Values) (interface{}, error) {
-	return nil, errNotImplemented
+func (h *handler) metaStats(entity *mongodoc.Entity, id *charm.Reference, path, method string, flags url.Values) (interface{}, error) {
+	req := charmstore.CounterRequest{
```
Comment (urosj) [inline internal/v4/api.go:312]: Could we retrieve number of downloads in some tome interval in the past and therefore does it make sense to add Start and Stop time to the CounterRequest?

File: internal/v4/api.go [hunk 4]
```diff
@@ -307,8 +308,20 @@ func (h *handler) metaArchiveSize(entity *mongodoc.Entity, id *charm.Reference,
...
+		Key: entityStatsKey(id, params.StatsArchiveDownload),
+	}
+	results, err := h.store.Counters(&req)
+	if err != nil {
+		return nil, errgo.Notef(err, "cannot retrieve stats")
+	}
+	return &params.StatsResponse{
+		// If a list is not requested as part of the charmstore.CounterRequest,
+		// one result is always returned: if the key is not found the count is
+		// set to 0.
```
Comment (urosj) [inline internal/v4/api.go:322]: So the return for nonexisting charm is 0 or error?

File: params/stats.go (hunk 1 above)
Comment (frankban) [inline params/stats.go:23]: I think so. Note that this struct pre-existed, I just moved it to stats.go.

File: internal/v4/stats.go (hunk 2 above)
Comment (frankban) [inline internal/v4/stats.go]: Done.

File: internal/v4/api.go (hunk 3 above)
Comment (frankban) [inline internal/v4/api.go:312]: I believe we can do that. This is just a start and it only considers absolute downloads.
We can refine what stats look like later I guess.

File: internal/v4/api.go (hunk 4 above)
Comment (frankban) [inline internal/v4/api.go:322]: The method is not even called for non-existing entities.
A not found is returned by the ResolveURL function passed to (and invoked by) the router.
0 is returned when there is no "archive-download" keys associated to the entity.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results: http://ci.jujugui.org:8080/job/charmstore/147/

**Comment by urosj** [comment]
:+1:  Thank you.

**Comment by frankban** [comment]
Thanks for the reviews!
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

**Pull request merged**